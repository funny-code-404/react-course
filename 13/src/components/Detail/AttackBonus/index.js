import PropTypes from "prop-types";

const AttackBonus = ({ data }) => (
	data && 
	<div>
		<p>Attack bonus:</p>
		<ul>
			{data.map((x) => <li key={x}>{x}</li>)}
		</ul>
	</div>
)

AttackBonus.propTypes = {
	data: PropTypes.arrayOf(PropTypes.string),
}

export default AttackBonus;