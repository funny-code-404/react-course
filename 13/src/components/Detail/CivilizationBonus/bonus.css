.civilization-bonus ul{
    padding: 1% 2%;
}
.civilization-bonus ul li{
    line-height: 2;
    text-transform: uppercase;
}