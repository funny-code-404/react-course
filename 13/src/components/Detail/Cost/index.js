import PropTypes from "prop-types";

const Cost = ({ data }) => (
	data && 
	<div>
		<p>Costs:</p>
		<ul>
			{Object.keys(data).map((e, i)=> (
				<li key={i} >{e}<span>{data[e]}</span></li>
			))}
		</ul>
	</div>
)

Cost.propTypes = {
	data: PropTypes.objectOf(PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
	])).isRequired,
}


export default Cost;