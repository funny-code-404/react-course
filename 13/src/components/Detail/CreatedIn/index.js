import React from "react";
import PropTypes from "prop-types";

import { CreatedIn } from "../../../pages";


class DetailCreatedIn extends React.Component {
    state = {
		created_in: false,
    }
    
    hendleCreatedIn = (e) => {
		this.setState(prevState => ({
			...prevState,
			created_in: !prevState.created_in,
		}))
    }
    
    render() {
        const { data } = this.props
        const result = data.split('/')[data.split('/').length - 1].replace('_', ' ');
        return(
            <div>
                <p>Created in:</p>
                        <button onClick={this.hendleCreatedIn}>
                            {
                                !this.state.created_in 
                                ? result
                                : 'close'
                            }
                        </button>
                        {
                            this.state.created_in && <CreatedIn url={data} />
                        }
            </div>
        )
    }
}

DetailCreatedIn.propTypes = {
    data: PropTypes.string.isRequired
}

export default DetailCreatedIn;