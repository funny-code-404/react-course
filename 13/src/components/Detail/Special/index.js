import PropTypes from "prop-types";

const Special = ({ data }) => (
	data && 
	<div>
		<p>Special:</p>
		<ul>
			{data.map((x) => <li key={x}>{x}</li>)}
		</ul>
	</div>
)

Special.propTypes = {
    data: PropTypes.arrayOf(PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.arrayOf(PropTypes.string)
	]))
}

export default Special;