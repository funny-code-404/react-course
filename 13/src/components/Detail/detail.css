.detail {
    padding: 2% 2% 0;
    border: 2px solid black;
    border-radius: 6px;
}
.detail p{
    line-height: 2;
    font-weight: 600;
    text-transform: uppercase;
}
.detail span{
    font-weight: 500;
}
.ditail-block{
    padding: 2% 0;
}

.detail button {
	border: none;
	border-radius: 0;
	color: white;
	background: black;
	padding: 5px 8px;
	text-transform: uppercase;
	font-size: 12px;
    cursor: pointer;
    outline: none; 
    margin-bottom: 1%;
}
