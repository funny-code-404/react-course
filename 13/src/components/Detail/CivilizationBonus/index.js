import PropTypes from "prop-types";

import './bonus.css';

const CivilizationBonus = ({ data }) => (
	data && 
	<div className='civilization-bonus'>
		<p>Civilization bonus:</p>
		<ul>
			{data.map((x) => <li key={x}>{x}</li>)}
		</ul>
	</div>
)

CivilizationBonus.propTypes = {
	data: PropTypes.arrayOf(PropTypes.string),
}

export default CivilizationBonus;