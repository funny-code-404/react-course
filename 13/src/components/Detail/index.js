import React from "react";
import PropTypes from "prop-types";

import "./detail.css";
import DetailCreatedIn from "./CreatedIn";
import DetailUniqueUnit from "./UniqueUnit";
import DetailUniqueTech from "./UniqueTech";
import CivilizationBonus from "./CivilizationBonus";
import AttackBonus from "./AttackBonus";
import Special from "./Special";
import Cost from "./Cost";

class Detail extends React.Component {
    render() {
        const {
            name,
            army_type,
            expansion,
            team_bonus,
            civilization_bonus,
            unique_unit,
            unique_tech,
            description,
            age,
            created_in,
            cost,
            build_time,
            reload_time,
            attack_delay,
            movement_rate,
            line_of_sight,
            hit_points,
            range,
            attack,
            armor,
            accuracy,
            special,
            attack_bonus,
            search_radius,
        } = this.props.data;

        const arrs = [
            { army_type },
            { team_bonus },
            { description },
            { expansion },
            { age },
            { build_time },
            { reload_time },
            { attack_delay },
            { movement_rate },
            { line_of_sight },
            { hit_points },
            { range },
            { attack },
            { armor },
            { search_radius },
            { accuracy },
        ];

        return (
            <div className="detail">
                <h1>{name}</h1>
                <div className="ditail-block">
                    {arrs.map((x) => {
                        for (const key in x) {
                            return (
                                x[key] && (
                                    <p key={key}>
                                        {key}: <span>{x[key]}</span>
                                    </p>
                                )
                            );
                        }
                    })}
                    {cost && <Cost data={cost} />}
                    {attack_bonus && <AttackBonus data={attack_bonus} />}
                    {special && <Special data={special} />}
                    {civilization_bonus && (
                        <CivilizationBonus data={civilization_bonus} />
                    )}
                    {unique_unit && Boolean(unique_unit.length) && (
                        <DetailUniqueUnit data={unique_unit} />
                    )}
                    {unique_tech && Boolean(unique_tech.length) && (
                        <DetailUniqueTech data={unique_tech} />
                    )}
                    {created_in && Boolean(created_in.length) && (
                        <DetailCreatedIn data={created_in} />
                    )}
                </div>
            </div>
        );
    }
}

Detail.propTypes = {
    data: PropTypes.objectOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
            PropTypes.arrayOf(PropTypes.string),
            PropTypes.objectOf(
                PropTypes.oneOfType([PropTypes.number, PropTypes.string])
            ),
        ])
    ),
};

export default Detail;
