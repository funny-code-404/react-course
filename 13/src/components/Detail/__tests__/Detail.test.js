import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import Detail from "..";
import DetailCreatedIn  from "../CreatedIn";
import DetailUniqueUnit  from "../UniqueUnit";
import DetailUniqueTech  from "../UniqueTech";
import CivilizationBonus from "../CivilizationBonus";
import AttackBonus from "../AttackBonus";
import Special from "../Special";
import Cost from "../Cost";

configure({ adapter: new Adapter() });

describe('DETAIL', () => {
    const props = {
        data: { 
            name: 'test', 
            army_type: 'text',
            civilization_bonus: 'test', 
            unique_unit: [ 'test' ], 
            unique_tech: [ 'test' ], 
            created_in: 'test'  , 
            cost: 'test', 
            special: 'test',
            attack_bonus: 'text', 
        },
    }

    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<Detail {...props} />);
    });

    it('to Match Snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('Component is return 1 content "h1" ', () => {
        expect(wrapper.find('h1')).toHaveLength(1);
    });
    it('Component is return 1 content "p" ', () => {
        expect(wrapper.find('p')).toHaveLength(1);
    });
    it('Component is return 1 content "span" ', () => {
        expect(wrapper.find('span')).toHaveLength(1);
    });
    it('Component is return 1 content "cost" ', () => {
        expect(wrapper.find(Cost)).toHaveLength(1)
    })
    it('Component is return 1 content "attack bonus" ', () => {
        expect(wrapper.find(AttackBonus)).toHaveLength(1)
    })
    it('Component is return 1 content "special" ', () => {
        expect(wrapper.find(Special)).toHaveLength(1)
    })
    it('Component is return 1 content "civilizations bonus" ', () => {
        expect(wrapper.find(CivilizationBonus)).toHaveLength(1)
    })
    it('Component is return 1 content "unique unit" ', () => {
        expect(wrapper.find(DetailUniqueUnit)).toHaveLength(1)
    })
    it('Component is return 1 content "unique tech" ', () => {
        expect(wrapper.find(DetailUniqueTech)).toHaveLength(1)
    })
    it('Component is return 1 content "created in" ', () => {
        expect(wrapper.find(DetailCreatedIn)).toHaveLength(1)
    })

    describe('Component should render correct content', () => {
        it('Component should render correct props in "h1" ', () => {
            expect(wrapper.find('h1').props().children).toEqual(props.data.name);
        });
        it('Component should render correct props in "span" ', () => {
            expect(wrapper.find('span').props().children).toEqual(props.data.army_type);
        });
        it('Component should render correct after props changing in "span" ', () => {
            const new_text = 'NEW';
            wrapper.setProps({ data: { army_type: new_text }})
            expect(wrapper.find('span').props().children).toEqual(new_text);
        });
        it('Component should render correct after props changing in "h1" ', () => {
            const new_text = 'NEW';
            wrapper.setProps({ data: { name: new_text }})
            expect(wrapper.find('h1').props().children).toEqual(new_text);
        });
    })
});

// npm run test src/components/Detail