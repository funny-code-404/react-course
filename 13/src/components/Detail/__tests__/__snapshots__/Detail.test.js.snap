// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DETAIL to Match Snapshot 1`] = `
<div
  className="detail"
>
  <h1>
    test
  </h1>
  <div
    className="ditail-block"
  >
    <p
      key="army_type"
    >
      army_type
      : 
      <span>
        text
      </span>
    </p>
    <Cost
      data="test"
    />
    <AttackBonus
      data="text"
    />
    <Special
      data="test"
    />
    <CivilizationBonus
      data="test"
    />
    <DetailUniqueUnit
      data={
        Array [
          "test",
        ]
      }
    />
    <DetailUniqueTech
      data={
        Array [
          "test",
        ]
      }
    />
    <DetailCreatedIn
      data="test"
    />
  </div>
</div>
`;
